shared-bindings/_eve/__init__.c

```diff
@@ -0,0 +1,1061 @@
+    mp_obj_t self = args[0];
+    mp_obj_t num = args[1];
+    mp_buffer_info_t fmt;
+    mp_get_buffer_raise(args[2], &fmt, MP_BUFFER_READ);
+    size_t len;
+    mp_obj_t *items;
+    mp_obj_tuple_get(args[3], &len, &items);
+
+    // Count how many 32-bit words required
+    size_t n = 0;
+    for (size_t i = 0; i < fmt.len; n++) {
+        switch (((char*)fmt.buf)[i]) {
+            case 'I':
+            case 'i':
+                i += 1;
+                break;
+            case 'H':
+            case 'h':
+                i += 2;
+                break;
+            default:
+                break;
+        }
+    }
+
+    uint32_t buf[16];
+    uint32_t *p = buf;
+    *p++ = 0xffffff00 | mp_obj_get_int_truncated(num);
+    mp_obj_t *a = items;
+    uint32_t lo;
+
+    for (size_t i = 0; i < fmt.len; ) {
+        switch (((char*)fmt.buf)[i]) {
+            case 'I':
+            case 'i':
+                *p++ = mp_obj_get_int_truncated(*a++);
+                i += 1;
+                break;
+            case 'H':
+            case 'h':
+                lo = mp_obj_get_int_truncated(*a++) & 0xffff;
+                *p++ = lo | (mp_obj_get_int_truncated(*a++) << 16);
+                i += 2;
+                break;
+            default:
+                break;
+        }
+    }
+
+    common_hal__eve_add(EVEHAL(self), sizeof(uint32_t) * (1 + n), buf);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(cmd_obj, 4, 4, _cmd);
+
+STATIC const mp_rom_map_elem_t _EVE_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_register), MP_ROM_PTR(&register_obj) },
+    { MP_ROM_QSTR(MP_QSTR_cc), MP_ROM_PTR(&cc_obj) },
+    { MP_ROM_QSTR(MP_QSTR_flush), MP_ROM_PTR(&flush_obj) },
+    { MP_ROM_QSTR(MP_QSTR_Vertex2f), MP_ROM_PTR(&vertex2f_obj) },
+    { MP_ROM_QSTR(MP_QSTR_cmd), MP_ROM_PTR(&cmd_obj) },
+    { MP_ROM_QSTR(MP_QSTR_cmd0), MP_ROM_PTR(&cmd0_obj) },
+    ROM_DECLS
+};
+STATIC MP_DEFINE_CONST_DICT(_EVE_locals_dict, _EVE_locals_dict_table);
+
+STATIC void _EVE_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    (void)self_in;
+    (void)kind;
+    mp_printf(print, "<_EVE>");
+}
+
+STATIC mp_obj_t _EVE_make_new(const mp_obj_type_t *type, size_t n_args, const mp_obj_t *args, mp_map_t *kw_args) {
+    // mp_arg_check_num(n_args, kw_args, 1, 1, false);
+    mp_obj__EVE_t *o = m_new_obj(mp_obj__EVE_t);
+    o->base.type = &_EVE_type;
+    o->_eve.n = 0;
+    o->_eve.vscale = 16;
+    return MP_OBJ_FROM_PTR(o);
+}
+
+STATIC const mp_obj_type_t _EVE_type = {
+    { &mp_type_type },
+    // Save on qstr's, reuse same as for module
+    .name = MP_QSTR__EVE,
+    .print = _EVE_print,
+    .make_new = _EVE_make_new,
+    // .attr = _EVE_attr,
+    .locals_dict = (void*)&_EVE_locals_dict,
+};
+
+STATIC const mp_rom_map_elem_t mp_module__eve_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR__eve) },
+    { MP_ROM_QSTR(MP_QSTR__EVE), MP_OBJ_FROM_PTR(&_EVE_type) },
```

Comment:
I'd drop the _ on the EVE class since the module name already has it. I'm ok if you'd prefer to keep it though.

Reply:
I'd prefer to keep it thanks. The two modules are used like this

```
from _eve import _EVE
from eve import EVE

class Gameduino(_EVE, EVE):
...
```

So it's a little more explicit to keep the underscore naming.

Reply:
Can you have EVE subclass _EVE instead? I'm a little nervous that the multi-inheritance isn't well tested.